### ISSUE TITLE
always crashed at cs_close() when full dynamic tracing used

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#1071]
Here is the execution environment.
```
$ uftrace --version
uftrace v0.9.3-189-gd9049 ( dwarf python luajit tui perf sched dynamic )

$ cat /etc/os-release | grep VERSION=
VERSION="18.04.3 LTS (Bionic Beaver)"
```
In this server, full-dynamic tracing always gets crashed as follows:
```
$ uftrace record -L. -P. --debug-domain dynamic:3 t-abc
dynamic: dynamic patch type: t-abc: 0 (none)
dynamic: patch normal func: a (patch size: 9)
dynamic: patch normal func: b (patch size: 9)
dynamic: patch normal func: c (patch size: 9)
dynamic: patch normal func: main (patch size: 8)
dynamic: patched all (4) functions in 't-abc'
WARN: child terminated by signal: 11: Segmentation fault
```
Here is the crash point.
```
$ cgdb -q --args uftrace record -L. -P. t-abc
    ...
 29│ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
 30│ {
 31├───────> cs_close(&disasm->engine);
 32│ }
 33│
 34│ enum fail_reason {
 35│         INSTRUMENT_FAIL_NODETAIL        = (1U << 0),
 36│         INSTRUMENT_FAIL_NOOPRND         = (1U << 1),
/home/honggyu/work/uftrace/arch/x86_64/mcount-insn.c
```
I deleted `cs_close` statement for debugging purpose.
```diff
diff --git a/arch/x86_64/mcount-insn.c b/arch/x86_64/mcount-insn.c
index 32c87c04..c1810489 100644
--- a/arch/x86_64/mcount-insn.c
+++ b/arch/x86_64/mcount-insn.c
@@ -28,7 +28,6 @@ void mcount_disasm_init(struct mcount_disasm_engine *disasm)

 void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
 {
-       cs_close(&disasm->engine);
 }

 enum fail_reason {
```
It still gets crashed in a different place.
```
1642│ static __used void mcount_startup(void)
1643│ {
1644│         char *pipefd_str;
1645│         char *logfd_str;
    ...
1769│         if (patch_str)
1770│                 mcount_dynamic_update(&symtabs, patch_str, patt_type, &disasm);
1771│
    ...
1781├───────> pthread_atfork(atfork_prepare_handler, NULL, atfork_child_handler);
1782│
1783│         mcount_hook_functions();
1784│
/home/honggyu/work/uftrace/libmcount/mcount.c
```

**Comment by honggyukim** [comment]
The backtrace of actual crash point is as follows:
```
(gdb) r
Starting program: /home/honggyu/work/uftrace/uftrace record -L. -P. t-abc
    ...
Thread 2.1 "t-abc" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff7fde840 (LWP 15368)]
do_lookup_x (undef_name=undef_name@entry=0x7ffff7b8eea6 "cs_close", new_hash=new_hash@entry=1382585776, old_hash=old_hash@entry=0x7fffffffcf80, ref=0x7ffff7b8dbe0,
    result=result@entry=0x7fffffffcf90, scope=<optimized out>, i=0, version=0x0, flags=1, skip=0x0, type_class=1, undef_map=0x7ffff7ff5270) at dl-lookup.c:385
385     dl-lookup.c: No such file or directory.

(gdb) bt
#0  do_lookup_x (undef_name=undef_name@entry=0x7ffff7b8eea6 "cs_close", new_hash=new_hash@entry=1382585776, old_hash=old_hash@entry=0x7fffffffcf80, ref=0x7ffff7b8dbe0,
    result=result@entry=0x7fffffffcf90, scope=<optimized out>, i=0, version=0x0, flags=1, skip=0x0, type_class=1, undef_map=0x7ffff7ff5270) at dl-lookup.c:385
#1  0x00007ffff7de01ef in _dl_lookup_symbol_x (undef_name=0x7ffff7b8eea6 "cs_close", undef_map=0x7ffff7ff5270, ref=ref@entry=0x7fffffffd028, symbol_scope=0x7ffff7ff55c8,
    version=0x0, type_class=type_class@entry=1, flags=1, skip_map=<optimized out>) at dl-lookup.c:813
#2  0x00007ffff7de4ec3 in _dl_fixup (l=<optimized out>, reloc_arg=<optimized out>) at ../elf/dl-runtime.c:112
#3  0x00007ffff7dec7ca in _dl_runtime_resolve_xsavec () at ../sysdeps/x86_64/dl-trampoline.h:125
#4  0x00007ffff7bbc399 in mcount_disasm_finish (disasm=0x7ffff7dd2c80 <disasm>) at /home/honggyu/work/uftrace/arch/x86_64/mcount-insn.c:31
#5  0x00007ffff7b9bf4e in finish_dynamic_update (disasm=0x7ffff7dd2c80 <disasm>) at /home/honggyu/work/uftrace/libmcount/dynamic.c:452
#6  0x00007ffff7b9c1e3 in mcount_dynamic_update (symtabs=0x7ffff7dd25a0 <symtabs>, patch_funcs=0x7fffffffeef3 ".", ptype=PATT_REGEX, disasm=0x7ffff7dd2c80 <disasm>)
    at /home/honggyu/work/uftrace/libmcount/dynamic.c:498
#7  0x00007ffff7b9797e in mcount_startup () at /home/honggyu/work/uftrace/libmcount/mcount.c:1770
#8  0x00007ffff7b97bb3 in mcount_init () at /home/honggyu/work/uftrace/libmcount/mcount.c:1877
#9  0x00007ffff7de5733 in call_init (env=0x7fffffffdce8, argv=0x7fffffffdcd8, argc=1, l=<optimized out>) at dl-init.c:72
#10 _dl_init (main_map=0x7ffff7ffe170, argc=1, argv=0x7fffffffdcd8, env=0x7fffffffdce8) at dl-init.c:119
#11 0x00007ffff7dd60ca in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#12 0x0000000000000001 in ?? ()
#13 0x00007fffffffdfdd in ?? ()
#14 0x0000000000000000 in ?? ()
```

**Comment by namhyung** [comment]
Did you do something with capstone?  As it's crashed in the dynamic linker, it'd not be a problem in uftrace or libmcount.  I'd recommend you re-install the capstone library..

FYI, I cannot reproduce with the current code:

```
$ ./uftrace -L. -P. --debug-domain dynamic:3 t-abc
dynamic: dynamic patch type: t-abc: 0 (none)
dynamic: patch normal func: a (patch size: 9)
dynamic: patch normal func: b (patch size: 9)
dynamic: patch normal func: c (patch size: 9)
dynamic: patch normal func: main (patch size: 8)
dynamic: patched all (4) functions in 't-abc'
# DURATION     TID     FUNCTION
            [ 43448] | main() {
            [ 43448] |   a() {
            [ 43448] |     b() {
            [ 43448] |       c() {
   1.799 us [ 43448] |         getpid();
   3.905 us [ 43448] |       } /* c */
   4.246 us [ 43448] |     } /* b */
   4.584 us [ 43448] |   } /* a */
   5.344 us [ 43448] | } /* main */
```

**Comment by honggyukim** [comment]
I have a different machine and capstone is installed by `misc/install-deps.sh`. It's only shown in a specific machine. I know it's weird but left the issue to track the issue in the future.

**Comment by nikita-uvarov** [comment]
Hello.

I am having exactly the same issue: full dynamic tracing in uftrace works normally on several machines (tested on ubuntu versions 16.04, 19.04), whereas crashes exactly as mentioned above on the other (happens to have ubuntu 18.04, not sure if related).

I've tried compiling uftrace statically (it complains some functions will be loaded from glibc dynamically, however, none of them are required for the launch above). The same binary works on "good" machines and crashes on "bad" one. So I believe the issue is not related to a particular libcapstone version.

Does anyone have any clues on what to look for?

**Comment by namhyung** [comment]
@nikita-uvarov thanks for the report.  Can you show me gdb backtrace of the crash in the static binary?

**Comment by honggyukim** [comment]
Hmm.. I've seen more and more often on the recently installed Ubuntu 18.04 servers.
```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

$ gcc tests/s-abc.c 

$ uftrace -P main --debug-domain dynamic:1 a.out 
dynamic: dynamic patch type: a.out: 0 (none)
dynamic: dynamic patch stats for 'a.out'
dynamic:    total:        1
dynamic:  patched:        0 ( 0.00%)
dynamic:   failed:        1 (100.00%)
dynamic:  skipped:        0 ( 0.00%)
dynamic: no match:        0
WARN: child terminated by signal: 11: Segmentation fault
WARN: cannot open record data: /tmp/uftrace-live-06hvuA: No data available
```

**Comment by honggyukim** [comment]
I tested it with `--no-libcall` again, but crashed as well.
```
$ uftrace -P main --debug-domain dynamic:1 --no-libcall a.out 
dynamic: dynamic patch type: a.out: 0 (none)
dynamic: dynamic patch stats for 'a.out'
dynamic:    total:        1
dynamic:  patched:        0 ( 0.00%)
dynamic:   failed:        1 (100.00%)
dynamic:  skipped:        0 ( 0.00%)
dynamic: no match:        0
WARN: child terminated by signal: 11: Segmentation fault
WARN: cannot open record data: /tmp/uftrace-live-qqpnY6: No data available
```
But the backtrace is a bit less deeper.
```
Thread 2.1 "a.out" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff7fcf840 (LWP 24651)]
do_lookup_x (undef_name=undef_name@entry=0x7ffff7b9472c "__register_atfork", new_hash=new_hash@entry=4201216622, old_hash=old_hash@entry=0x7fd140, ref=0x7ffff7b934d8, result=result@entry=0x7fd
150, scope=<optimized out>, i=0, version=0x7ffff7fd4c98, flags=1, skip=0x0, type_class=1, undef_map=0x7ffff7ff5030) at dl-lookup.c:385
385     dl-lookup.c: No such file or directory.
(gdb) bt
#0  do_lookup_x (undef_name=undef_name@entry=0x7ffff7b9472c "__register_atfork", new_hash=new_hash@entry=4201216622, old_hash=old_hash@entry=0x7fd140, ref=0x7ffff7b934d8, result=result@entry=      0x7ffff
fffd150, scope=<optimized out>, i=0, version=0x7ffff7fd4c98, flags=1, skip=0x0, type_class=1, undef_map=0x7ffff7ff5030) at dl-lookup.c:385
#1  0x00007ffff7de01ef in _dl_lookup_symbol_x (undef_name=0x7ffff7b9472c "__register_atfork", undef_map=0x7ffff7ff5030, ref=ref@entry=0x7fd1e8, symbol_scope=0x7ffff7ff5388, version=0x7ffff7fd4c98      , t
ype_class=type_class@entry=1, flags=1, skip_map=<optimized out>) at dl-lookup.c:813
#2  0x00007ffff7de4ec3 in _dl_fixup (l=<optimized out>, reloc_arg=<optimized out>) at ../elf/dl-runtime.c:112
#3  0x00007ffff7dec7ca in _dl_runtime_resolve_xsavec () at ../sysdeps/x86_64/dl-trampoline.h:125
#4  0x00007ffff7b9c3a4 in mcount_startup () at /home/honggyu/work/uftrace/libmcount/mcount.c:1781
#5  0x00007ffff7b9c55e in mcount_init () at /home/honggyu/work/uftrace/libmcount/mcount.c:1877
#6  0x00007ffff7de5733 in call_init (env=0x7fddf8, argv=0x7fdde8, argc=1, l=<optimized out>)             at dl-init.c:72
#7  _dl_init (main_map=0x7ffff7ffe170, argc=1, argv=0x7fdde8, env=      0x7fddf8) at dl-init.c:119
#8  0x00007ffff7dd60ca in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#9  0x01 in ?? ()
#10 0x00007fe12f in ?? ()
#11 0x0 in ?? ()
```

**Comment by honggyukim** [comment]
@ParkHanbum Could you please take a look at it if you can reproduce the problem?

**Comment by honggyukim** [comment]
I see the same problem in another Ubuntu 16.04 machine as well.
```
$ cat /etc/os-release | grep VERSION=
VERSION="16.04.6 LTS (Xenial Xerus)"
```

**Comment by honggyukim** [comment]
Here is another backtrace with `--no-libcall`
```
(gdb) bt
#0  do_lookup_x (undef_name=undef_name@entry=0x7ffff7b8fed6 "cs_close", new_hash=new_hash@entry=1382585776, old_hash=old_hash@entry=0x7fffffffcfe
0, ref=0x7ffff7b8ec10, result=result@entry=0x7fffffffcff0, scope=<optimized out>, i=0, version=0x0, flags=1, skip=0x0, type_class=1, undef_map=0x
7ffff7ff6000) at dl-lookup.c:402
#1  0x00007ffff7de1b1f in _dl_lookup_symbol_x (undef_name=0x7ffff7b8fed6 "cs_close", undef_map=0x7ffff7ff6000, ref=ref@entry=0x7fffffffd0a8, symb
ol_scope=0x7ffff7ff6358, version=0x0, type_class=type_class@entry=1, flags=1, skip_map=0x0) at dl-lookup.c:829
#2  0x00007ffff7de6ad6 in _dl_fixup (l=<optimized out>, reloc_arg=<optimized out>) at ../elf/dl-runtime.c:111
#3  0x00007ffff7deef5a in _dl_runtime_resolve_xsavec () at ../sysdeps/x86_64/dl-trampoline.h:129
#4  0x00007ffff7bbd423 in mcount_disasm_finish (disasm=0x7ffff7dd4c80 <disasm>) at /home/honggyu/uftrace/arch/x86_64/mcount-ins
n.c:31
#5  0x00007ffff7b99da7 in finish_dynamic_update (disasm=0x7ffff7dd4c80 <disasm>) at /home/honggyu/uftrace/libmcount/dynamic.c:4
52
#6  0x00007ffff7b9a03c in mcount_dynamic_update (symtabs=0x7ffff7dd45a0 <symtabs>, patch_funcs=0x7fffffffeeea "main", ptype=PATT_REGEX, disasm=0x
7ffff7dd4c80 <disasm>) at /home/honggyu/uftrace/libmcount/dynamic.c:498
#7  0x00007ffff7b989e4 in mcount_startup () at /home/honggyu/uftrace/libmcount/mcount.c:1776
#8  0x00007ffff7b98c19 in mcount_init () at /home/honggyu/uftrace/libmcount/mcount.c:1883
#9  0x00007ffff7de76ca in call_init (l=<optimized out>, argc=argc@entry=1, argv=argv@entry=0x7fffffffdd78, env=env@entry=0x7fffffffdd88) at dl-in
it.c:72
#10 0x00007ffff7de77db in call_init (env=0x7fffffffdd88, argv=0x7fffffffdd78, argc=1, l=<optimized out>) at dl-init.c:30
#11 _dl_init (main_map=0x7ffff7ffe168, argc=1, argv=0x7fffffffdd78, env=0x7fffffffdd88) at dl-init.c:120
#12 0x00007ffff7dd7c6a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#13 0x0000000000000001 in ?? ()
#14 0x00007fffffffe0bd in ?? ()
#15 0x0000000000000000 in ?? ()
```

**Comment by honggyukim** [comment]
It also gets crashed even if capstone is disabled.
```
$ ./configure --without-capstone

$ make DEBUG=1 install

$ gcc -o t-abc tests/s-abc.c

$ uftrace record -P main --no-libcall t-abc
WARN: child terminated by signal: 11: Segmentation fault
```
The backtrace is a bit different.
```
(gdb) bt
#0  do_lookup_x (undef_name=undef_name@entry=0x7ffff7b91ea6 "__register_atfork", new_hash=new_hash@entry=4201216622, old_hash=old_hash@entry=0x7f
ffffffd0a0, ref=0x7ffff7b908e0, result=result@entry=0x7fffffffd0b0, scope=<optimized out>, i=0, version=0x7ffff7fd5938, flags=1, skip=0x0, type_c
lass=1, undef_map=0x7ffff7ff6000) at dl-lookup.c:402
#1  0x00007ffff7de1b1f in _dl_lookup_symbol_x (undef_name=0x7ffff7b91ea6 "__register_atfork", undef_map=0x7ffff7ff6000, ref=ref@entry=0x7fffffffd
168, symbol_scope=0x7ffff7ff6358, version=0x7ffff7fd5938, type_class=type_class@entry=1, flags=1, skip_map=0x0) at dl-lookup.c:829
#2  0x00007ffff7de6ad6 in _dl_fixup (l=<optimized out>, reloc_arg=<optimized out>) at ../elf/dl-runtime.c:111
#3  0x00007ffff7deef5a in _dl_runtime_resolve_xsavec () at ../sysdeps/x86_64/dl-trampoline.h:129
#4  0x00007ffff7b9a8bf in mcount_startup () at /home/honggyu/uftrace/libmcount/mcount.c:1787
#5  0x00007ffff7b9aa79 in mcount_init () at /home/honggyu/uftrace/libmcount/mcount.c:1883
#6  0x00007ffff7de76ca in call_init (l=<optimized out>, argc=argc@entry=1, argv=argv@entry=0x7fffffffdd78, env=env@entry=0x7fffffffdd88) at dl-in
it.c:72
#7  0x00007ffff7de77db in call_init (env=0x7fffffffdd88, argv=0x7fffffffdd78, argc=1, l=<optimized out>) at dl-init.c:30
#8  _dl_init (main_map=0x7ffff7ffe168, argc=1, argv=0x7fffffffdd78, env=0x7fffffffdd88) at dl-init.c:120
#9  0x00007ffff7dd7c6a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#10 0x0000000000000001 in ?? ()
#11 0x00007fffffffe0bd in ?? ()
#12 0x0000000000000000 in ?? ()
```
The crashed location is as follows.
```c
1648│ static __used void mcount_startup(void)
1649│ {
...
1784│         if (getenv("UFTRACE_KERNEL_PID_UPDATE"))
1785│                 kernel_pid_update = true;
1786│
1787├───────> pthread_atfork(atfork_prepare_handler, NULL, atfork_child_handler);
1788│
1789│         mcount_hook_functions();
1790│
1791│         /* initialize script binding */
1792│         if (SCRIPT_ENABLED && script_str)
1793│                 mcount_script_init(patt_type);
1794│
1795│         compiler_barrier();
1796│         pr_dbg("mcount setup done\n");
1797│
1798│         mcount_global_flags &= ~MCOUNT_GFL_SETUP;
1799│         mtd.recursion_marker = false;
1800│ }
```

**Comment by honggyukim** [comment]
It seems that the problem is related to the following `mprotect` in `./arch/x86_64/mcount-dynamic.c`.
```c
void mcount_cleanup_trampoline(struct mcount_dynamic_info *mdi)
{
        if (mprotect(PAGE_ADDR(mdi->text_addr),
                         PAGE_LEN(mdi->text_addr, mdi->text_size), PROT_EXEC))
                pr_err("cannot restore trampoline due to protection");
}
```

**Comment by honggyukim** [comment]
It works fine after applying the following change.
```diff
diff --git a/arch/x86_64/mcount-dynamic.c b/arch/x86_64/mcount-dynamic.c
index 57b3e5f..0175e4a 100644
--- a/arch/x86_64/mcount-dynamic.c
+++ b/arch/x86_64/mcount-dynamic.c
@@ -125,9 +125,6 @@ int mcount_setup_trampoline(struct mcount_dynamic_info *mdi)

 void mcount_cleanup_trampoline(struct mcount_dynamic_info *mdi)
 {
-       if (mprotect(PAGE_ADDR(mdi->text_addr),
-                        PAGE_LEN(mdi->text_addr, mdi->text_size), PROT_EXEC))
-               pr_err("cannot restore trampoline due to protection");
 }

 static void read_xray_map(struct arch_dynamic_info *adi,
```
```
$ uftrace -P. --no-libcall t-abc
# DURATION     TID     FUNCTION
            [ 50405] | main() {
            [ 50405] |   a() {
            [ 50405] |     b() {
   2.060 us [ 50405] |       c();
   4.087 us [ 50405] |     } /* b */
   5.149 us [ 50405] |   } /* a */
   6.956 us [ 50405] | } /* main */
```
We may have to fix the above `mprotect` properly.

**Comment by honggyukim** [comment]
Hmm.. The following change makes it work, but I don't fully understand why we have to give read permission as well.
```diff
diff --git a/arch/x86_64/mcount-dynamic.c b/arch/x86_64/mcount-dynamic.c
index 57b3e5f..43d8a44 100644
--- a/arch/x86_64/mcount-dynamic.c
+++ b/arch/x86_64/mcount-dynamic.c
@@ -126,7 +126,7 @@ int mcount_setup_trampoline(struct mcount_dynamic_info *mdi)
 void mcount_cleanup_trampoline(struct mcount_dynamic_info *mdi)
 {
        if (mprotect(PAGE_ADDR(mdi->text_addr),
-                        PAGE_LEN(mdi->text_addr, mdi->text_size), PROT_EXEC))
+                        PAGE_LEN(mdi->text_addr, mdi->text_size), PROT_READ|PROT_EXEC))
                pr_err("cannot restore trampoline due to protection");
 }

```
@AnsBal @namhyung @ParkHanbum Do you have any idea on that?

**Comment by honggyukim** [comment]
@nikita-uvarov Can you also verify if the above change fixes your problem as well?

**Comment by namhyung** [comment]
It probably has some RO data along with the code.

**Comment by honggyukim** [comment]
If so, I’m also wondering why it was okay with other *good* machines.

**Comment by namhyung** [comment]
It's tricky since it depends on how the code and rodata is located which might vary due to a different combination of cpu/compiler/kernel versions.

**Comment by AnsBal** [comment]
I couldn't reproduce the bug and throw a look at the .text section to find out where the read is done. If you would like to share the compiled binary (t-abc), I would be happy to investigate it.
In general, in /proc/self/maps, the memory region of the main executable has always the read permission (in my system at least):
 ```
anas@anas-GE72-2QD:~$ cat /proc/self/maps 
56505c9ec000-56505c9f4000 r-xp 00000000 08:06 131096                     /bin/cat
56505cbf3000-56505cbf4000 r--p 00007000 08:06 131096                     /bin/cat
56505cbf4000-56505cbf5000 rw-p 00008000 08:06 131096                     /bin/cat
...
```
I'm not surprised the READ permission solved the problem :)

**Comment by honggyukim** [comment]
@AnsBal Thanks for checking it. I compiled `t-abc` on the previously crashed machine, then send it to another machine, which didn't have this problem.

Even without the commit I made, `t-abc` from the crashed machine doesn't get crashed in another machine. So you will not see the problem even if I share the binary.

I think it's tricky to fully understand the problem, but having READ_PROT along with EXEC_PROT seems to be normally used as you mentioned.
